api/python/quilt3/backends/base.py

```diff
@@ -0,0 +1,174 @@
+    def resolve_top_hash(self, pkg_name: str, hash_prefix: str) -> str:
+        pass
+
+    @abc.abstractmethod
+    def shorten_top_hash(self, pkg_name: str, top_hash: str) -> str:
+        pass
+
+
+class PackageRegistryV1(PackageRegistry):
+    root_path = '.quilt'
+
+    @property
+    def pointers_global_dir(self) -> PhysicalKey:
+        return self.root.join('named_packages/')
+
+    def pointers_dir(self, pkg_name: str) -> PhysicalKey:
+        return self.root.join(f'named_packages/{pkg_name}/')
+
+    def pointer_pk(self, pkg_name: str, pointer_name: str) -> PhysicalKey:
+        return self.root.join(f'named_packages/{pkg_name}/{pointer_name}')
+
+    def pointer_latest_pk(self, pkg_name: str) -> PhysicalKey:
+        return self.pointer_pk(pkg_name, self.latest_tag_name)
+
+    @property
+    def manifests_global_dir(self) -> PhysicalKey:
+        return self.root.join('packages/')
+
+    def manifests_package_dir(self, pkg_name) -> PhysicalKey:
+        return self.manifests_global_dir  # TODO: does this make sense?
+
+    def manifest_pk(self, pkg_name: str, top_hash: str) -> PhysicalKey:
+        return self.root.join(f'packages/{top_hash}')
+
+    def delete_package(self, pkg_name: str, top_hash: str = None):
+        package_path = self.pointers_dir(pkg_name)
+        paths = list(list_url(package_path))
+        if not paths:
+            raise QuiltException("No such package exists in the given directory.")
+
+        if top_hash is not None:
+            top_hash = self.resolve_top_hash(pkg_name, top_hash)
+            deleted = []
+            remaining = []
+            for path, pkg_hash in self.list_package_versions(pkg_name):
+                (deleted if pkg_hash == top_hash else remaining).append(path)
+            if not deleted:
+                raise QuiltException("No such package version exists in the given directory.")
+            for path in deleted:
+                delete_url(package_path.join(path))
+            if 'latest' in deleted and remaining:
+                # Create a new "latest". Technically, we need to compare numerically,
+                # but string comparisons will be fine till year 2286.
+                new_latest = max(remaining)
+                copy_file(package_path.join(new_latest), package_path.join('latest'))
+        else:
+            for path, _ in paths:
+                delete_url(package_path.join(path))
+
+    def push_manifest(self, pkg_name: str, top_hash: str, manifest_data: bytes):
+        put_bytes(manifest_data, self.manifest_pk(pkg_name, top_hash))
+        hash_bytes = top_hash.encode()
+        named_path = self.pointers_dir(pkg_name)
+        # TODO: use a float to string formatter instead of double casting
+        timestamp_path = named_path.join(str(int(time.time())))
+        put_bytes(hash_bytes, timestamp_path)
+        put_bytes(hash_bytes, self.pointer_latest_pk(pkg_name))
+
+    def resolve_top_hash(self, pkg_name: str, hash_prefix: str) -> str:
```

Comment:
To support the existing API, don't we need a way to resolve short hashes across all package names?

Reply:
If we decide to keep `Package.resolve_hash()` public we should add `pkg_name` parameter for it. It would be optional for V1 registry and required for V2 registry.

Reply:
Let's deprecate and make private in V2.